chain/chain/src/validate.rs

```diff
@@ -18,41 +22,97 @@ use crate::{ChainStore, Error, ErrorKind, RuntimeAdapter};
 const GAS_LIMIT_ADJUSTMENT_FACTOR: u64 = 1000;
 
 /// Verifies that chunk's proofs in the header match the body.
-pub fn validate_chunk_proofs(
-    chunk: &ShardChunk,
-    runtime_adapter: &dyn RuntimeAdapter,
-) -> Result<bool, Error> {
+pub fn validate_chunk_proofs(chunk: &ShardChunk, runtime_adapter: &dyn RuntimeAdapter) -> bool {
     // 1. Checking chunk.header.hash
-    if chunk.header.hash != ChunkHash(hash(&chunk.header.inner.try_to_vec()?)) {
+    if chunk.header.hash != ChunkHash(hash(&chunk.header.inner.try_to_vec().unwrap())) {
         byzantine_assert!(false);
-        return Ok(false);
+        return false;
     }
 
     // 2. Checking that chunk body is valid
     // 2a. Checking chunk hash
     if chunk.chunk_hash != chunk.header.hash {
         byzantine_assert!(false);
-        return Ok(false);
+        return false;
     }
     // 2b. Checking that chunk transactions are valid
     let (tx_root, _) = merklize(&chunk.transactions);
     if tx_root != chunk.header.inner.tx_root {
         byzantine_assert!(false);
-        return Ok(false);
+        return false;
     }
     // 2c. Checking that chunk receipts are valid
     if chunk.header.inner.height_created == 0 {
-        return Ok(chunk.receipts.len() == 0
-            && chunk.header.inner.outgoing_receipts_root == CryptoHash::default());
+        return chunk.receipts.len() == 0
+            && chunk.header.inner.outgoing_receipts_root == CryptoHash::default();
     } else {
-        let outgoing_receipts_hashes = runtime_adapter.build_receipts_hashes(&chunk.receipts)?;
+        let outgoing_receipts_hashes = runtime_adapter.build_receipts_hashes(&chunk.receipts);
         let (receipts_root, _) = merklize(&outgoing_receipts_hashes);
         if receipts_root != chunk.header.inner.outgoing_receipts_root {
             byzantine_assert!(false);
-            return Ok(false);
+            return false;
         }
     }
-    Ok(true)
+    true
+}
+
+/// Validates transactions in the given chunk. Checks that the given transactions are in proper
+/// order and also checks that every transaction are in the same chain as the given block header.
+/// Returns true if chunk transactions are valid.
+pub fn validate_chunk_transactions(
+    chain_store: &mut ChainStore,
+    transactions: &[SignedTransaction],
+    block_header: &BlockHeader,
+    transaction_validity_period: BlockIndex,
+) -> bool {
+    if !validate_transactions_order(transactions) {
+        return false;
+    }
+
+    // Verifying transactions are on the same chain as the block header.
+    let all_transactions_are_valid = transactions.iter().all(|t| {
```

Comment:
This check passes in two scenarios: when the hash of the transaction is before the hash of the block on the chain, and when the hash of the transaction is after (or equal) the hash of the block on the chain. Shouldn't only the latter be allowed?

Reply:
I didn't touch `check_blocks_on_same_chain` code. So not sure if this is the case @bowenwang1996

Reply:
I think it only passes when in the latter case.

Reply:
@bowenwang1996 Could you please point to the code (specific lines) where we can see that it passes only in the latter case?

Reply:
https://github.com/nearprotocol/nearcore/blob/f93e13c0be52b8069a98de4792b861f07787f003/chain/chain/src/store.rs#L401

Reply:
I mean within the function?

Reply:
This is unrelated to this PR. Created the issue to track it: https://github.com/nearprotocol/nearcore/issues/1836